Assets/Modules/Movement/Scripts/PlayerMovementController.cs

```diff
@@ -0,0 +1,70 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.InputSystem;
+
+public class PlayerMovementController : MonoBehaviour
+{
+
+    [SerializeField] private float jumpForce = 350f;
+
+    private bool gamepadIsConnected = false;
+
+    private Camera mainCamera;
+    private Rigidbody2D rb;
+    private MyInputActions characterActions;
+
+    private Vector2 aim;
+
+
+    // Start is called before the first frame update
+    void Start()
+    {
+        rb = GetComponent<Rigidbody2D>();
+        mainCamera = Camera.main;
+        characterActions = new MyInputActions();
+        characterActions.Enable();
+
+        // Checking if Gamepad is connected. Only does it at Start, so if gamepad is off, start play and THEN turn on gamepad, it will jump towards mouse.
+        // If starts with gamepad, then jumps with mouse, it explodes xD
+        foreach(var device in InputSystem.devices){
```

Comment:
You should be able to listen to the device change in the input system and then run this loop whenever a new device is connected/disconnected:
Maybe in awake or start subscribe to the device change event:
```
 InputSystem.onDeviceChange += OnDeviceChange;
```
and then
```
private void OnDeviceChange(InputDevice device, InputDeviceChange change)
    {
        // Update input devices when a device is added or removed
        if (change == InputDeviceChange.Added || change == InputDeviceChange.Removed)
        {
            // run the loop
        }
    }
```

Reply:
What does "subscribing to the device change event" mean? Is it like saying to the program "I want to hear from you in case the device list changes"?

Then I create that method "OnDeviceChange" that will make sure that only when 1 device is added or removed I run the loop, correct?
Thanks!

Reply:
Also do I need the "InputDevice device" as a parameter of this OnDeviceChange method? because inside the "foreach" loop I'm already declaring that. I'm actually getting an error:
![Screenshot 2024-06-19 at 09 41 17](https://github.com/SwiffterAnim/Bounce-Escape/assets/170499803/5d6932bb-81b1-4f14-b553-8021ffd24758)

Reply:
Sorry, another image for clarification:
![Screenshot 2024-06-19 at 09 42 55](https://github.com/SwiffterAnim/Bounce-Escape/assets/170499803/c3ee45cc-b18d-49ac-b2f8-3ca4c0da431a)

Reply:
Deleting the InputDevice device from the parenthesis of the method, or changing the "var device" fixes the issue.

Reply:
> What does "subscribing to the device change event" mean? Is it like saying to the program "I want to hear from you in case the device list changes"?
> 
> Then I create that method "OnDeviceChange" that will make sure that only when 1 device is added or removed I run the loop, correct? Thanks!

You got the subscribing concept right. In this line
```
InputSystem.onDeviceChange += OnDeviceChange;
```
the += is like "please fire this function whenever a device changed.

Reply:
to make it more pretty you can even do this one line only:
```
gamepadIsConnected = device is Joystick;
```

Reply:
That's interesting about the subscribing! I didn't know about that concept! pretty cool!
And cool, I fixed the note.

There are a couple things in the movement that I'd like to make sure we do, but I'll let you decide the order of things ok?
1- I want him to not be able to double or triple jump. On my original, once he jumps, he can only jump again if he hits the wall, or if he connects into a hook, or if he gets hit by something.
2- I want the force/velocity to reset every time we jump, right now, if we jump to one side, then jump to the other side, it's like the physics of the rigidbody are adding or subtracting force from the already moving object. Meaning, if we jump to the left and then to the right in quick succession, right now the end result would be that he ends up still on the screen, rather than jumping to the right. I think I fixed this issue before by resetting the physics somehow on my prototype.

I'll also adjust the force of the jump and the gravity later on.

Reply:
> That's interesting about the subscribing! I didn't know about that concept! pretty cool! And cool, I fixed the note.
> 
> There are a couple things in the movement that I'd like to make sure we do, but I'll let you decide the order of things ok? 1- I want him to not be able to double or triple jump. On my original, once he jumps, he can only jump again if he hits the wall, or if he connects into a hook, or if he gets hit by something. 2- I want the force/velocity to reset every time we jump, right now, if we jump to one side, then jump to the other side, it's like the physics of the rigidbody are adding or subtracting force from the already moving object. Meaning, if we jump to the left and then to the right in quick succession, right now the end result would be that he ends up still on the screen, rather than jumping to the right. I think I fixed this issue before by resetting the physics somehow on my prototype.
> 
> I'll also adjust the force of the jump and the gravity later on.

Regarding point 1, you can easily do that using colliders and checking the collision in code (I guess that's how you did it last time) also you need to have a flag that shows if the player has jumped and locks the jump action until it hits the wall of a hook.
As for point 2, I guess you can also set the velocity to zero as soon as the collision is detected.
Since we still don't have the Obstacles implemented (hook/wall) just use the famous gameObject.CompareTag("") and check the tag "Obstacle" for now to identify collisions. If you don't know which type of collision you should check  and also how to setup the collider components, let me know 👍

Reply:
Ok, I fixed both these issues!
One other issue that I'm having now, that I don't remember having on my prototype, is why doesn't the character bounce OFF the walls? I'm going against the wall and sticking to it, rather than bouncing off of it.. I would like to bounce off of it, like a pool ball against the walls of the pool table =')
I'm also having difficulties with the Mouse VS Gamepad thing, it's bugging completely. So I commented out on my latest version the possibility of aiming with the mouse and it's working better. We can try figuring this out later. I'm losing too much time with that =')

Reply:
hmm I think you need to create a `PhysicsMaterial` and assign it to the wall's Rigidbody. Then you can adjust the bounciness on the material. If it didn't work try to assign it to the player rigidbody as well.

Reply:
Yes that's what it was! Thanks hehe